onnxruntime/contrib_ops/cuda/collective/sharded_moe.h

```diff
@@ -24,10 +24,17 @@ class ShardedMoE final : public NcclKernel, public MoEBase {
 
  private:
   Status SynchronizeExpertsStartIndex(AllocatorPtr& alloc, OpKernelContext* ctx, cudaEvent_t& cuda_event) const;
+  Status SynchronizeExpertsStartIndexImpl(AllocatorPtr& alloc, OpKernelContext* ctx, cudaEvent_t& cuda_event) const;
 
   int64_t local_experts_start_index_;
   int64_t tensor_shards_;
-  std::vector<int64_t> rank_to_experts_start_index_;
+  mutable InlinedVector<int64_t> rank_to_experts_start_index_;
+
+  // A global resource pack for IPC memory used in custom reduce kernel.
+  // Resource retrieval and deserialization are made atomic to thread safety of accessing it.
+  mutable onnxruntime::cuda::collective::GlobalIPCMemoryResourcePack g_ipc_mem_res_pack_;
```

Comment:
nit: It is global instance, there is no need to add it as member variable.
      You can access the instance by IPCMemoryResourcePack::GetGlobalInstance() as mentioned in another comment.